File: Linked List/109.convert-sorted-list-to-binary-search-tree.py
#
# @lc app=leetcode id=109 lang=python3
#
# [109] Convert Sorted List to Binary Search Tree
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def findMiddle(self, head):
        slow, fast, prev = head, head, None
        while fast and fast.next:
            prev = slow
            slow = slow.next
            fast = fast.next.next
        if prev:   # cut the link
            prev.next  = None
        return slow

    def sortedListToBST(self, head):
        mid = self.findMiddle(head)
        if not head:
            return None
        node  = TreeNode(mid.val)
        if head == mid:   # one element in linked list 
            return node
        node.left = self.sortedListToBST(head)
        node.right = self.sortedListToBST(mid.next)
        return node
            # @lc code=end

File: Linked List/109.convert-sorted-list-to-binary-search-tree.cpp
/*
 * @lc app=leetcode id=109 lang=cpp
 *
 * [109] Convert Sorted List to Binary Search Tree
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* findMiddle(ListNode* head)
    {
        auto fast = head;
        auto slow = head;
        ListNode* prev = NULL;
        while(fast && fast->next)
        {
            prev = slow;
            slow = slow->next;
            fast = fast->next->next;
        }
        if(prev) prev->next = NULL;
        return slow;
    }
    TreeNode* sortedListToBST(ListNode* head)
    {
        if(!head) return NULL;
        auto mid = findMiddle(head);
        auto node = new TreeNode(mid->val);
        if(head == mid) return node;
        node->left = sortedListToBST(head);
        node->right = sortedListToBST(mid->next);
        return node;
    }
};

// @lc code=end

